### PR TITLE
feat(table filter): add custom filter function

### PR DESCRIPTION
**Pull request by YardWill**
### 👀 PR includes

✨ Feature

- [x] New feature

🎨 Enhance

- [ ] Code style optimization
- [ ] Refactoring
- [ ] Change the UI
- [ ] Improve the performance
- [ ] Type optimization

🐛 Bugfix

- [ ] Solve the issue and close #0

🔧 Chore

- [ ] Test case
- [ ] Docs / demos update
- [ ] CI / workflow
- [ ] Release version
- [ ] Other ()

### 📝 Description

add custom filter function feat

### 🖼️ Screenshot

| Before | After |
| ------ | ----- |
| ❌      | ✅     |

### 🔍 Self Check before Merge

- [ ] Add or update relevant Docs.
- [ ] Add or update relevant Demos.
- [ ] Add or update relevant TypeScript definitions.

### CODE REVIEW

File: packages/s2-core/src/common/interface/basic.ts [hunk 1]
```diff
@@ -141,7 +140,8 @@ export interface SortParam extends Sort {
 
 export interface FilterParam {
   filterKey: string;
-  filteredValues: unknown[];
+  filteredValues?: unknown[];
+  filterFunction?: (row: DataType) => boolean;
```
Comment (lijinke666) [inline packages/s2-core/src/common/interface/basic.ts:144]: filterFunction 这个名字 太抽象了吧...

就叫 `filter` ? 或者 `customFilter`

File: packages/s2-core/src/data-set/table-data-set.ts [hunk 2]
```diff
@@ -56,13 +57,12 @@ export class TableDataSet extends BaseDataSet {
...
+    each(this.filterParams, ({ filterKey, filteredValues, customFilter }) => {
+      const defaultFilterFunc = (row: DataType) =>
+        row[filterKey] && !includes(filteredValues, row[filterKey]);
       this.displayData = [
         ...this.getStartRows(),
-        ...filter(
-          this.getMovableRows(),
-          (row) => row[filterKey] && !includes(filteredValues, row[filterKey]),
-        ),
+        ...filter(this.getMovableRows(), customFilter || defaultFilterFunc),
```
Comment (serializedowen) [inline packages/s2-core/src/data-set/table-data-set.ts:65]: filteredValues and customFilter are supposed to be independent, with neither taking precedence over each other

**Review by serializedowen — CHANGES_REQUESTED** [review]

**Review by zxc0328 — APPROVED** [review]
LG

**Review by serializedowen — APPROVED** [review]

**Pull request merged**